src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,190 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import logging
+import os
+from collections.abc import Iterable
+from functools import wraps
+from pathlib import Path
+from typing import Callable, Literal, TypeVar
+
+import tomlkit
+from tomlkit import TOMLDocument
+from tomlkit.items import Table
+
+from ..errors import ConfigParserError, ConfigSourceError
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ConfigOption:
+    def __init__(
+        self,
+        name: str,
+        _root_parser: ConfigParser,
+        _nest_path: list[str],
+        _type: Callable[[str], _T] | None = None,
+        choices: Iterable[_T] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: The name of the ConfigOption
+            env_name: Environmental variable value should be read from, if not supplied, we'll construct this
+            type: A function that can turn str to the desired type, useful for reading value from environmental variable
+        """
+        self.name = name
+        self.type = _type
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigParser = _root_parser
+        self.env_name = env_name
+
+    def get(self):
+        """Retrieve a value of option."""
+        source = "environment variable"
+        loaded_env, value = self._get_env()
+        if not loaded_env:
+            source = "configuration file"
+            value = self._get_config()
+        if self.choices and value not in self.choices:
+            raise ConfigSourceError(
+                f"The value of {self.option_name} read from "
+                f"{source} is not part of {self.choices}"
+            )
+        return value
+
+    @property
+    def option_name(self) -> str:
+        return ".".join(self._nest_path[1:])
+
+    @property
+    def option_env_name(self) -> str:
+        pieces = map(lambda e: e.upper(), self._nest_path[1:])
+        return f"SF{'_' + '_'.join(pieces)}"
```

Comment:
I propose that we use the shorthand `SF`

Reply:
but in our design doc, we intentionally decided to use full name to avoid confusion of SF. Snowservice and SnowML have already used SNOWFLAKE prefix. We should align with the original decision. https://docs.google.com/document/d/1mi3mciW7AQkWiGozhC8edzX0S_PNonf5fty4YIIOgfA/edit#heading=h.uy91oh2ef1m1

Reply:
After some more discussion, I updated the PR to use `SNOWFLAKE_` as its default prefix